### PR TITLE
Initial attempt to add palette support

### PR DESCRIPTION
**Pull request by lwabeke**
Allows saving of files using palette table
Example usage below:

```
using PNGFiles, Images

nclasses = 15
myArray = [UInt8.(mod(x+y, nclasses)) for x=1:640, y=1:480 ]
PNGFiles.save("myfile_pal.png", myArray, palette=colortable)

```

### CODE REVIEW

**Comment by Drvi** [comment]
Thanks for looking into this! I was just working on supporting 1, 2 and 4 `bit_depths` for grayscale which could be, iiuc, extended to paletted images as well.

**Comment by lwabeke** [comment]
This was added in response to this question:
https://discourse.julialang.org/t/saving-2d-arrays-as-indexed-color-images/36399/12

**Comment by lwabeke** [comment]
But probably needs more work to make sure it doesn't break something, like selecting an invalid bit_depth

**Comment by Drvi** [comment]
@lwabeke, since this feature is really useful and you laid down the groundwork, I took the liberty and finished it. Currently, we could only support 8-bit palettes; 1, 2 and 4-bit depths need a little bit more packing work to happen.

One thing I'm not happy about: Currently, when you could provide a vector of `RBA{N0f8}` or `RGBA{N0f8}` and the `image` is basically a matrix of _zero based indices_ to it. I restrict the `eltype` of the image to be `UInt8` and explicitly mention the zero-based indexing it in the docstring, but I still feel this can get confusing for users.

cc: @ianshmean

**Comment by lwabeke** [comment]
Thanks, I took a few stabs in the dark and it seemed very promising, but you took it a lot further.

Yes, the zero-based indexing could be confusing since Julia is mostly 1 based. I guess you could try to autodetect it, if min(image) == 0, assume zero-based, if max(image) == length(palette), assume 1-based, then correct offset. But you are left with the potential inconsistent/weird behaviour if extrema of image doesn't go to the palette edges.
You could alternatively try to use a custom data type/types instead of UInt8 (zero-based and one-based indexing types), but that feels like it is going to cause other hassles for users to cast their image data into these new types. Unless you use it as a very thin wrapper (i.e syntactic sugar to help user understanding):
```
mutable struct ZeroBasedIndexing
          image :: AbstractArray
end

mutable struct OneBasedIndexing
          image :: AbstractArray
end

PNGFiles.save("myfile_pal.png", ZeroBasedIndexing.(myArray), palette=colortable)
PNGFiles.save("myfile_pal.png", OneBasedIndexing.(myArray), palette=colortable)
```
with the convert function of OneBasedIndexing applying the -1 offset correction, something like
```
Base.convert(::Type{ZeroBasedIndexing}, oneBased :: OneBasedIndexing) = ZeroBasedIndexing(oneBased.image.-1)

PNGFiles.save(fname, oneBased :: OneBasedIndexing, palette=colortable) = PNGFiles.save(fname, ZeroBasedIndexing(oneBased.image.-1), palette=colortable)
```

I'm not sure if you need the Base.convert or if there is a way to get Julia to automatically call the convert, without having to provide the extra specialisation to save

**Comment by timholy** [comment]
I haven't reviewed this, but just a brief comment to consider leveraging [IndirectArrays](https://github.com/JuliaArrays/IndirectArrays.jl) (which is a tiny package) and [OffsetArrays](https://github.com/JuliaArrays/OffsetArrays.jl) for the zero-based indexing.

**Comment by Drvi** [comment]
Thanks for the ideas, @timholy @lwabeke. I think that using `IndirectArrays` and manually changing its `index` is the way to go.

**Comment by ianshmean** [comment]
Sounds like a good strategy. @Drvi I think you have a much better handle on this than me. Feel free to merge when you think it's ready. I'd just suggest that we also check impact on package load time, but I can't imagine that would change much

**Comment by ianshmean** [comment]
Seems like travis's windows isn't letting go of the test files before the cleanup happens. 

Given clean-up isn't critical on CI, perhaps just disable the cleanup on CI

https://github.com/lwabeke/PNGFiles.jl/blob/30daf9fcb88474b914825a1e7720b5022f5605b1/test/runtests.jl#L53

```julia
function is_ci()
    get(ENV, "TRAVIS", "") == "true" ||
    get(ENV, "APPVEYOR", "") in ("true", "True") ||
    get(ENV, "CI", "") in ("true", "True")
end

if !is_ci()
    # Cleanup
    isdir(PNG_TEST_PATH) && rm(PNG_TEST_PATH, recursive = true)
end
```

**Comment by Drvi** [comment]
Just fyi, I'm working on this. For symmetry, I'd like to allow reading paletted PNGs as IndirectArrays with OffsetArray palette, but there are some cases in PNGSuite images that give me a hard time.

File: src/io.jl [hunk 1]
```diff
@@ -114,7 +148,14 @@ function load(fpath::String; gamma::Union{Nothing,Float64}=nothing)
...
     png_destroy_read_struct(Ref{Ptr{Cvoid}}(png_ptr), Ref{Ptr{Cvoid}}(info_ptr), C_NULL)
     close_png(fp)
-    return transpose(buffer)
+    buffer = permutedims(buffer, (2, 1))
+    if expand_paletted || color_type != PNG_COLOR_TYPE_PALETTE
+        return buffer
+    else
+        # We got 0-based indices back from libpng and converting to 1-based could overflow UInt8.
+        # Using UInt16 for index would cost us large part of the savings provided by IndirectArray.
+        return IndirectArray(buffer, OffsetArray(palette, 0:length(palette)-1))
```
Comment (johnnychen94) [inline src/io.jl:157]: I think this is equivalent to `palette[buffer .+ 1]` but not sure if they have the same performance.

Comment (Drvi) [reply src/io.jl:157]: Ah, that is more elegant, thanks!

**Comment by Drvi** [comment]
Not sure what is going on with the FreeBSD, but I'm gonna merge this so we can move forward. Thanks everybody!

**Comment by ianshmean** [comment]
Awesome. Just an afterthought, but is this worth a read/write example in the readme?

**Pull request merged**